### PR TITLE
Implement nonlinear spring force actuator

### PR DESCRIPTION
**Pull request by eh111eh**
#### References to other Issues or PRs

PR #25518

#### Brief description of what is fixed or changed
The Duffing equation is a nonlinear second order differential equation used to model the dynamics of systems, damped and driven oscillators, with a cubic nonlinearity in the restoring force, which is given by $$\\ddot{x} + \delta\dot{x} + \beta x + \alpha x^3 = \gamma \cos(\omega t)\$$

However, as a full Duffing second order forced system will not work as an actuator, I modify the Duffing Spring to treat it as a non-second order system.
For $\beta>0$, the Duffing oscillator can be interpreted as a forced oscillator with a spring whose restoring force is written as $$F = -\beta x - \alpha x^3$$
Since the case  $\beta<0$ causes the chaotic motions, I will not consider this case for now. 

I Implemented a new class `DuffingSpring` to model the force exerted by a nonlinear spring, based on the Duffing equation: $F = -\beta x - \alpha x^3$
where $x$ is the displacement from the equilibrium position, $\beta$ is the linear spring constant, and $\alpha$ is the coefficient for the nonlinear cubic term.

#### Other comments
**Note**: This pull request has been moved to [#26438](https://github.com/sympy/sympy/pull/26438). Please refer to the new pull request for further updates and discussions.

Reference
- http://www.scholarpedia.org/article/Duffing_oscillator 
- https://www.cfm.brown.edu/people/dobrush/am34/Matlab/ch3/duffing.html
- https://www.cfm.brown.edu/people/dobrush/am34/Mathematica/ch3/duffing.html

#### Release Notes

* physics.mechanics
  * implemented `DuffingSpring` to model the force exerted by a nonlinear spring, based on the Duffing equation

### CODE REVIEW

**Comment by moorepants** [comment]
It seems you have edited the wrong file here.

I don't think making a spring that simply k*f(x) will be that useful and you can manually do such a thing with little effort. It would be better to find some specific models of nonlinear springs that are in common use and implement it. I'm not suggesting to implement this spring but here is a spring model that is non-trivial to create: https://www.hindawi.com/journals/sv/2020/8814332/ (as an example).

**Comment by moorepants** [comment]
Here is an eardrum model: https://www.cfm.brown.edu/people/dobrush/am33/Mathematica/ch4/nspring.html

**Comment by moorepants** [comment]
This is probably a good candidate: https://en.wikipedia.org/wiki/Duffing_equation This is a widely  used spring actuator model.

**Comment by eh111eh** [comment]
> This is probably a good candidate: https://en.wikipedia.org/wiki/Duffing_equation This is a widely used spring actuator model.

Thanks for the comment. I've started working on the actuator for the Duffing equation, and this time I've updated the correct file.

File: sympy/physics/mechanics/actuator.py [hunk 1]
```diff
@@ -873,3 +882,109 @@ def __repr__(self):
...
+    """An actuator modeled by the Duffing equation.
+
+    Explanation
+    ===========
+
+    The Duffing equation is a non-linear differential equation used to model
+    various physical systems, especially oscillators with non-linear stiffness.
+    In its standard form, the equation is given by:
+
+        (d**2 * x) / (dt**2) + δ * (dx / dt) + α * x + β * x**3 = γ * cos(ω * t)
```
Comment (moorepants) [inline sympy/physics/mechanics/actuator.py:896]: I don't think we designed Acutators to be second order systems. You can make a nonlinear Duffing Spring, but I don't think a full Duffing second order forced system will work as a actuator.

File: sympy/physics/mechanics/actuator.py [hunk 2]
```diff
@@ -873,3 +882,109 @@ def __repr__(self):
...
+            t_eval = np.linspace(0, t_max, int(t_max * omega * points_per_period / (2 * np.pi)))
+            sol = solve_ivp(duffing_ode, [0, t_max], [x0, x_dot0], t_eval=t_eval)
+            x = sol.y[0]
+
+            # Extract the amplitude of the steady-state response.
+            steady_state = x[-len(x)//4:]
+            amplitude = (np.max(steady_state) - np.min(steady_state)) / 2
+            amplitudes.append(amplitude)
+
+        return np.array(amplitudes)
```
Comment (moorepants) [inline sympy/physics/mechanics/actuator.py:990]: None of this makes sense as an acuator. Have a look at how a spring actuator works and then try to make a Duffing Spring.

**Comment by eh111eh** [comment]
Hi! I modified the Duffing Spring to treat it as a non-second order system.
For $\beta>0$, the Duffing oscillator can be interpreted as a forced oscillator with a spring whose restoring force is written as $$F = -\beta x - \alpha x^3$$
Since the case  $\beta<0$ causes the chaotic motions, I will not consider this case for now. 

Reference
- http://www.scholarpedia.org/article/Duffing_oscillator

**Comment by moorepants** [comment]
Ok, nice. You will need to add a unit test that ensures your new class works as expected. See the other unit tests for actuators as examples.

I will also be help to see if this force works as expected. Can you demonstrate it on something simple like: https://pydy.readthedocs.io/en/latest/examples/mass-spring-damper.html ?

**Comment by moorepants** [comment]
It seems you edited the "facts.py" file and accidentally committed it here. I think the best solution would be for you to revert that change and then force push your branch back to Github. But an easier solution would be to just open a new pull request with only the Duffing actuator changes.

**Comment by oscarbenjamin** [comment]
> The following commits **delete files**:

It is not sufficient to just add the files back in subsequent commits. The files must not be deleted in the git history.

**Comment by eh111eh** [comment]
> > The following commits **delete files**:
> 
> It is not sufficient to just add the files back in subsequent commits. The files must not be deleted in the git history.

Sorry, I tried to revert the changes I made in facts.py and push those revisions. But it now seems better to create a new pull request with only the change in the Duffing actuator. Thank you for pointing this out!

**Pull request closed without merging**